gen-ai/orchestrator-server/src/main/python/server/tests/services/test_qa_chain.py

```diff
@@ -0,0 +1,132 @@
+#   Copyright (C) 2024 Credit Mutuel Arkea
+#
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+#
+from unittest.mock import AsyncMock, patch
+
+import pytest
+from langchain_core.documents import Document
+
+from gen_ai_orchestrator.models.vector_stores.vectore_store_provider import (
+    VectorStoreProvider,
+)
+from gen_ai_orchestrator.routers.requests.requests import QAQuery
+from gen_ai_orchestrator.services.langchain.qa_chain import execute_qa_chain
+
+
+# "Mock an item where it is used, not where it came from."
+# (https://www.toptal.com/python/an-introduction-to-mocking-in-python)
+# See https://docs.python.org/3/library/unittest.mock.html#where-to-patch
+# Here:
+# --> Not where it came from:
+# @patch("llm_orchestrator.services.langchain.factories.langchain_factory.get_llm_factory")
+# --> But where it is used (in the execute_qa_chain method of the llm_orchestrator.services.langchain.rag_chain
+# module that imports get_llm_factory):
+@patch('gen_ai_orchestrator.services.langchain.qa_chain.get_em_factory')
+@patch('gen_ai_orchestrator.services.langchain.qa_chain.get_vector_store_factory')
+@patch('gen_ai_orchestrator.services.langchain.qa_chain.QAResponse')
+@patch('gen_ai_orchestrator.services.langchain.qa_chain.build_chain')
+@patch('gen_ai_orchestrator.services.langchain.qa_chain.Source')
+@pytest.mark.asyncio
+async def test_qa_chain(
+    mocked_source,
+    mocked_build_chain,
+    mocked_qa_response,
+    mocked_get_vector_store_factory,
+    mocked_get_em_factory,
+):
+    """Test the full execute_qa_chain method by mocking all external calls."""
+    # Build a test RagQuery
+    query_dict = {
+        'history': [
```

Comment:
History isn't used in the chain, user might thinks that it's handled if we are giving an exemple with it.

Reply:
Yes, it's an oversight on my part.